rust-symcrypt/src/rsa.rs

```diff
@@ -0,0 +1,578 @@
+    fn store_msb_first_u64(&self, value: u64) -> Result<Vec<u8>, SymCryptError> {
+        let exp_size = self.size_of_public_exponent();
+        let mut dst = vec![0u8; exp_size as usize]; // Allocate tight size in bytes for storing public exponent
+        unsafe {
+            // SAFETY: FFI calls
+            match symcrypt_sys::SymCryptStoreMsbFirstUint64(
+                value,
+                dst.as_mut_ptr(),
+                exp_size as u64,
+            ) {
+                symcrypt_sys::SYMCRYPT_ERROR_SYMCRYPT_NO_ERROR => Ok(dst),
+                err => Err(SymCryptError::from(err)),
+            }
+        }
+    }
+
+    /// Load a `u64` from a byte slice assuming big-endian order.
+    fn load_msb_first_u64(src: &[u8]) -> Result<u64, SymCryptError> {
+        let mut dst = 0u64;
+        unsafe {
+            // SAFETY: FFI calls
+            match symcrypt_sys::SymCryptLoadMsbFirstUint64(
+                src.as_ptr(),
+                src.len() as u64,
+                &mut dst as *mut u64,
+            ) {
+                symcrypt_sys::SYMCRYPT_ERROR_SYMCRYPT_NO_ERROR => Ok(dst),
+                err => Err(SymCryptError::from(err)),
+            }
+        }
+    }
+}
+
+impl RsaKey for RsaKeyPair {
+    fn inner(&self) -> symcrypt_sys::PSYMCRYPT_RSAKEY {
+        self.inner
+    }
+
+    fn key_usage(&self) -> RsaKeyUsage {
+        self.key_usage
+    }
+}
+
+impl RsaKey for RsaPublicKey {
+    fn inner(&self) -> symcrypt_sys::PSYMCRYPT_RSAKEY {
+        self.inner
+    }
+
+    fn key_usage(&self) -> RsaKeyUsage {
+        self.key_usage
+    }
+}
+
+impl RsaKeyPair {
+    /// `generate_new()` generates a random Rsa key based on the provided parameters.
+    ///
+    /// `n_bits_mod` represents the desired bit length of the Rsa Key.
+    ///
+    /// `pub_exp` takes in an `Option<u64>` that is the public exponent. If `None` is provided, the default `2^16 +1` will be used.
+    ///  
+    /// `key_usage` will indicate if this key will be used for [`RsaKeyUsage::Sign`] or [`RsaKeyUsage::Encrypt`]
+    pub fn generate_new(
+        n_bits_mod: u32,
+        pub_exp: Option<&[u8]>,
+        key_usage: RsaKeyUsage,
+        num_format: NumberFormat,
+    ) -> Result<Self, SymCryptError> {
+        let rsa_key = allocate_rsa(2, n_bits_mod); // !Review: @Phil has mentioned that there is no scenario that we'd want to ONLY generate a public key, @Crypto folks is this the case?
+        let u64_pub_exp = match pub_exp {
+            Some(exp) => RsaKeyPair::load_msb_first_u64(exp)?,
+            None => DEFAULT_PUBLIC_EXPONENT, // if no public exponent is provided, use the default ( 2^16 + 1 )
+        };
+        unsafe {
+            // SAFETY: FFI calls
+            // No flags specified for SymCryptRsakeyAllocate
+            match symcrypt_sys::SymCryptRsakeyGenerate(
+                rsa_key,
+                [u64_pub_exp].as_ptr(), 
+                1,
```

Comment:
Might want to add a comment that this is an array of length 1.

Reply:
Updated.